File: Sources/SilknetDS/Components/SilknetButton.swift
// SilknetButton — SwiftUI port of @silknet-ds/react's Button.
// Verified against Figma node 4104:5038 (Button frame) and the React
// implementation in packages/react/src/Button/Button.css.
//
// API:
//
//     SilknetButton("Save") { /* onTap */ }
//     SilknetButton("Cancel", variant: .secondary)
//     SilknetButton("Add", variant: .primary, size: .lg, leftIcon: AnyView(Image(systemName: "plus")))
//     SilknetButton("Read more", variant: .link, size: .xs)
//
// Use the standard `.disabled(true)` modifier to disable.

import SwiftUI

public enum SilknetButtonVariant: Sendable {
    case primary
    case primarySoft
    case secondary
    case ghost
    case link
}

public enum SilknetButtonSize: Sendable {
    case xs
    case sm
    case md
    case lg
}

public struct SilknetButton: View {
    private let title: String
    private let variant: SilknetButtonVariant
    private let size: SilknetButtonSize
    private let leftIcon: AnyView?
    private let rightIcon: AnyView?
    private let action: () -> Void

    public init(
        _ title: String,
        variant: SilknetButtonVariant = .primary,
        size: SilknetButtonSize = .md,
        leftIcon: AnyView? = nil,
        rightIcon: AnyView? = nil,
        action: @escaping () -> Void = {}
    ) {
        self.title = title
        self.variant = variant
        self.size = size
        self.leftIcon = leftIcon
        self.rightIcon = rightIcon
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: gap) {
                if let leftIcon { leftIcon.frame(width: iconSize, height: iconSize) }
                Text(title)
                    .font(font)
                    .tracking(0.2)
                if let rightIcon { rightIcon.frame(width: iconSize, height: iconSize) }
            }
        }
        .buttonStyle(SilknetButtonStyle(variant: variant, size: size))
    }

    // MARK: - Size-derived layout values

    private var gap: CGFloat {
        switch size {
        case .xs, .sm: return .silknet.digitsSpacing2 // 8
        case .md, .lg: return .silknet.digitsSpacing3 // 12
        }
    }

    private var iconSize: CGFloat {
        switch size {
        case .xs, .sm: return 16
        case .md, .lg: return 20
        }
    }

    private var font: Font {
        switch size {
        case .xs, .sm: return .silknet.buttonButtonAdditional
        case .md, .lg: return .silknet.buttonButtonDefault
        }
    }
}

// MARK: - ButtonStyle (handles bg, border, padding, pressed overlay, disabled)

private struct SilknetButtonStyle: ButtonStyle {
    let variant: SilknetButtonVariant
    let size: SilknetButtonSize

    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        let pressed = configuration.isPressed
        let shape = RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)

        return configuration.label
            .padding(.horizontal, horizontalPadding)
            .padding(.vertical, verticalPadding)
            .foregroundColor(foregroundColor(pressed: pressed))
            .background(
                ZStack {
                    backgroundColor.clipShape(shape)
                    // Alpha-black overlay mirrors React's hover/pressed gradient
                    // (8% on hover, 16% on press). iOS has no hover — show only
                    // the pressed overlay. Link variant has no background to
                    // overlay; its color shifts via foregroundColor() above.
                    if pressed && isEnabled && variant != .link {
                        Color.black.opacity(0.16).clipShape(shape)
                    }
                }
            )
            .overlay(borderOverlay(shape: shape))
            .modifier(LinkUnderline(active: variant == .link && pressed && isEnabled))
            .contentShape(shape)
            .opacity(isEnabled ? 1 : 1) // disabled colors are explicit; keep opacity 1
    }

    // MARK: - Size

    private var horizontalPadding: CGFloat {
        if variant == .link { return 0 }
        switch size {
        case .xs: return .silknet.digitsSpacing2 // 8
        case .sm: return .silknet.digitsSpacing3 // 12
        case .md: return .silknet.digitsSpacing4 // 16
        case .lg: return .silknet.digitsSpacing5 // 20
        }
    }

    private var verticalPadding: CGFloat {
        if variant == .link { return 0 }
        switch size {
        case .xs: return .silknet.digitsSpacing1 // 4
        case .sm: return .silknet.digitsSpacing2 // 8
        case .md: return .silknet.digitsSpacing3 // 12
        case .lg: return .silknet.digitsSpacing4 // 16
        }
    }

    private var cornerRadius: CGFloat {
        switch size {
        case .xs, .sm: return .silknet.digitsRadiusXs // 8
        case .md, .lg: return .silknet.digitsRadiusS  // 12
        }
    }

    // MARK: - Colors

    private var backgroundColor: Color {
        // Disabled — variant-specific; matches React .silk-button:disabled rules.
        if !isEnabled {
            switch variant {
            case .secondary, .ghost, .link: return .clear
            case .primary, .primarySoft:    return .silknet.backgroundDisabled
            }
        }
        switch variant {
        case .primary:     return .silknet.backgroundPrimaryAccent
        case .primarySoft: return .silknet.backgroundPrimarySoft
        case .secondary:   return .silknet.backgroundLayer
        case .ghost, .link: return .clear
        }
    }

    private func foregroundColor(pressed: Bool) -> Color {
        // Disabled
        if !isEnabled {
            switch variant {
            case .primary, .primarySoft: return .silknet.textContrast  // white on grey — intentional ghost-out
            case .secondary, .ghost, .link: return .silknet.textDisabled
            }
        }
        // Enabled
        switch variant {
        case .primary: return .silknet.textContrast
        case .primarySoft: return .silknet.textPrimary
        case .secondary, .ghost: return .silknet.textSecondary
        case .link:
            // color-mix(primary, 8%/16% black) — approximated via opacity.
            // 1 → primary, 0.92 → hover-ish, 0.84 → pressed-ish.
            if pressed {
                return Color.silknet.backgroundPrimaryAccent.opacity(0.84)
            }
            return .silknet.backgroundPrimaryAccent
        }
    }

    @ViewBuilder
    private func borderOverlay(shape: RoundedRectangle) -> some View {
        if !isEnabled, variant == .secondary {
            shape.strokeBorder(Color.silknet.backgroundDisabled, lineWidth: 1)
        } else if isEnabled, variant == .secondary {
            shape.strokeBorder(Color.silknet.borderDefault, lineWidth: 1)
        }
    }
}

// Link variant: underline only when pressed (matches React's text-decoration
// rule). Pulled into its own ViewModifier so we can toggle without restructuring.
private struct LinkUnderline: ViewModifier {
    let active: Bool
    func body(content: Content) -> some View {
        if active {
            content.underline()
        } else {
            content
        }
    }
}

// MARK: - Xcode Previews

#if DEBUG
private struct PreviewSwatch<Content: View>: View {
    let label: String
    @ViewBuilder var content: Content
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(label)
                .font(.caption.monospaced())
                .foregroundColor(.secondary)
            content
        }
    }
}

#Preview("All variants — md") {
    VStack(alignment: .leading, spacing: 16) {
        PreviewSwatch(label: "primary")     { SilknetButton("Button", variant: .primary) }
        PreviewSwatch(label: "primary-soft"){ SilknetButton("Button", variant: .primarySoft) }
        PreviewSwatch(label: "secondary")   { SilknetButton("Button", variant: .secondary) }
        PreviewSwatch(label: "ghost")       { SilknetButton("Button", variant: .ghost) }
        PreviewSwatch(label: "link (xs)")   { SilknetButton("Button", variant: .link, size: .xs) }
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}

#Preview("Sizes — primary") {
    VStack(alignment: .leading, spacing: 16) {
        PreviewSwatch(label: "xs") { SilknetButton("Button", size: .xs) }
        PreviewSwatch(label: "sm") { SilknetButton("Button", size: .sm) }
        PreviewSwatch(label: "md") { SilknetButton("Button", size: .md) }
        PreviewSwatch(label: "lg") { SilknetButton("Button", size: .lg) }
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}

#Preview("Disabled") {
    VStack(alignment: .leading, spacing: 16) {
        PreviewSwatch(label: "primary disabled")     { SilknetButton("Button", variant: .primary).disabled(true) }
        PreviewSwatch(label: "primary-soft disabled"){ SilknetButton("Button", variant: .primarySoft).disabled(true) }
        PreviewSwatch(label: "secondary disabled")   { SilknetButton("Button", variant: .secondary).disabled(true) }
        PreviewSwatch(label: "ghost disabled")       { SilknetButton("Button", variant: .ghost).disabled(true) }
        PreviewSwatch(label: "link disabled")        { SilknetButton("Button", variant: .link, size: .xs).disabled(true) }
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
}

#Preview("Dark mode — all variants") {
    VStack(alignment: .leading, spacing: 16) {
        SilknetButton("Button", variant: .primary)
        SilknetButton("Button", variant: .primarySoft)
        SilknetButton("Button", variant: .secondary)
        SilknetButton("Button", variant: .ghost)
        SilknetButton("Button", variant: .link, size: .xs)
    }
    .padding(24)
    .background(Color.silknet.backgroundLayer)
    .preferredColorScheme(.dark)
}
#endif

File: android/library/src/main/kotlin/ge/silknet/ds/components/SilknetButton.kt
// SilknetButton — Jetpack Compose mirror of the iOS SilknetButton.
// Verified against Figma Button component (node 27:110 family).
// 5 variants × 4 sizes; pressed = 16% alpha-black overlay (matches React +
// SwiftUI overlay pattern).

package ge.silknet.ds.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import ge.silknet.ds.Silknet
import ge.silknet.ds.SilknetTheme

enum class SilknetButtonVariant { Primary, PrimarySoft, Secondary, Ghost, Link }
enum class SilknetButtonSize { Xs, Sm, Md, Lg }

@Composable
fun SilknetButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: SilknetButtonVariant = SilknetButtonVariant.Primary,
    size: SilknetButtonSize = SilknetButtonSize.Md,
    enabled: Boolean = true,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()

    val bg = backgroundFor(variant, enabled)
    val fg = foregroundFor(variant, enabled)
    val borderColor = borderFor(variant, enabled)
    val shape = RoundedCornerShape(cornerRadiusFor(size))

    val baseModifier = modifier
        .height(heightFor(size))
        .clip(shape)
        .background(bg)
        .let { if (borderColor != null) it.border(1.dp, borderColor, shape) else it }
        .clickable(
            interactionSource = interactionSource,
            indication = null,
            enabled = enabled,
            onClick = onClick,
        )

    Box(
        modifier = baseModifier,
        contentAlignment = Alignment.Center,
    ) {
        // Pressed overlay (skipped for link — link uses opacity darken instead).
        val pressedOverlayAlpha = when {
            !isPressed || !enabled -> 0f
            variant == SilknetButtonVariant.Link -> 0f
            else -> 0.16f
        }
        Box(
            modifier = Modifier
                .matchParentSize()
                .background(Color.Black.copy(alpha = pressedOverlayAlpha))
        )
        val effectiveFg = if (variant == SilknetButtonVariant.Link && isPressed && enabled) {
            fg.copy(alpha = 0.84f)
        } else fg

        Text(
            text = text,
            color = effectiveFg,
            style = textStyleFor(size, variant),
            textDecoration = if (variant == SilknetButtonVariant.Link)
                TextDecoration.Underline else TextDecoration.None,
            modifier = Modifier.padding(horizontal = horizontalPaddingFor(size, variant)),
        )
    }
}

// ---- style resolvers ----

@Composable
private fun backgroundFor(variant: SilknetButtonVariant, enabled: Boolean): Color {
    if (!enabled) {
        return when (variant) {
            SilknetButtonVariant.Primary, SilknetButtonVariant.PrimarySoft -> Silknet.colors.backgroundDisabled
            SilknetButtonVariant.Secondary, SilknetButtonVariant.Ghost, SilknetButtonVariant.Link -> Color.Transparent
        }
    }
    return when (variant) {
        SilknetButtonVariant.Primary -> Silknet.colors.backgroundPrimaryAccent
        SilknetButtonVariant.PrimarySoft -> Silknet.colors.backgroundPrimarySoft
        SilknetButtonVariant.Secondary -> Silknet.colors.backgroundLayer
        SilknetButtonVariant.Ghost, SilknetButtonVariant.Link -> Color.Transparent
    }
}

@Composable
private fun foregroundFor(variant: SilknetButtonVariant, enabled: Boolean): Color {
    if (!enabled) {
        return when (variant) {
            // Filled disabled = white text "ghosted out" (design intent — text-contrast on disabled-grey bg).
            SilknetButtonVariant.Primary, SilknetButtonVariant.PrimarySoft -> Silknet.colors.textContrast
            SilknetButtonVariant.Secondary, SilknetButtonVariant.Ghost, SilknetButtonVariant.Link -> Silknet.colors.textDisabled
        }
    }
    return when (variant) {
        SilknetButtonVariant.Primary -> Silknet.colors.textContrast
        SilknetButtonVariant.PrimarySoft -> Silknet.colors.textPrimary
        SilknetButtonVariant.Secondary, SilknetButtonVariant.Ghost -> Silknet.colors.textSecondary
        SilknetButtonVariant.Link -> Silknet.colors.textPrimary
    }
}

@Composable
private fun borderFor(variant: SilknetButtonVariant, enabled: Boolean): Color? {
    if (variant != SilknetButtonVariant.Secondary) return null
    return if (enabled) Silknet.colors.borderDefault else Silknet.colors.backgroundDisabled
}

private fun heightFor(size: SilknetButtonSize): Dp = when (size) {
    SilknetButtonSize.Xs -> 24.dp
    SilknetButtonSize.Sm -> 32.dp
    SilknetButtonSize.Md -> 44.dp
    SilknetButtonSize.Lg -> 52.dp
}

@Composable
private fun cornerRadiusFor(size: SilknetButtonSize): Dp = when (size) {
    SilknetButtonSize.Xs, SilknetButtonSize.Sm -> Silknet.dimens.digitsRadiusXs
    SilknetButtonSize.Md, SilknetButtonSize.Lg -> Silknet.dimens.digitsRadiusS
}

@Composable
private fun horizontalPaddingFor(size: SilknetButtonSize, variant: SilknetButtonVariant): Dp {
    if (variant == SilknetButtonVariant.Link) return 0.dp
    return when (size) {
        SilknetButtonSize.Xs -> Silknet.dimens.digitsSpacing2
        SilknetButtonSize.Sm -> Silknet.dimens.digitsSpacing3
        SilknetButtonSize.Md -> Silknet.dimens.digitsSpacing4
        SilknetButtonSize.Lg -> Silknet.dimens.digitsSpacing5
    }
}

@Composable
private fun textStyleFor(size: SilknetButtonSize, variant: SilknetButtonVariant): TextStyle {
    val base = when (size) {
        SilknetButtonSize.Xs, SilknetButtonSize.Sm -> Silknet.typography.buttonButtonAdditional
        SilknetButtonSize.Md, SilknetButtonSize.Lg -> Silknet.typography.buttonButtonDefault
    }
    return if (variant == SilknetButtonVariant.Link) {
        when (size) {
            SilknetButtonSize.Xs, SilknetButtonSize.Sm -> Silknet.typography.buttonLinkbuttonAdditional
            SilknetButtonSize.Md, SilknetButtonSize.Lg -> Silknet.typography.buttonLinkbuttonDefault
        }
    } else base
}

// ---- previews ----

@Preview(name = "All variants — md")
@Composable
private fun PreviewAllVariants() {
    SilknetTheme {
        Box(
            modifier = Modifier
                .background(Silknet.colors.backgroundLayer)
                .padding(24.dp)
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    SilknetButton("Primary", {}, variant = SilknetButtonVariant.Primary)
                    SilknetButton("Soft", {}, variant = SilknetButtonVariant.PrimarySoft)
                    SilknetButton("Secondary", {}, variant = SilknetButtonVariant.Secondary)
                }
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    SilknetButton("Ghost", {}, variant = SilknetButtonVariant.Ghost)
                    SilknetButton("Link", {}, variant = SilknetButtonVariant.Link)
                }
            }
        }
    }
}

@Preview(name = "Sizes — primary")
@Composable
private fun PreviewSizes() {
    SilknetTheme {
        Box(
            modifier = Modifier
                .background(Silknet.colors.backgroundLayer)
                .padding(24.dp)
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                SilknetButton("xs", {}, size = SilknetButtonSize.Xs)
                SilknetButton("sm", {}, size = SilknetButtonSize.Sm)
                SilknetButton("md", {}, size = SilknetButtonSize.Md)
                SilknetButton("lg", {}, size = SilknetButtonSize.Lg)
            }
        }
    }
}

@Preview(name = "Disabled")
@Composable
private fun PreviewDisabled() {
    SilknetTheme {
        Box(
            modifier = Modifier
                .background(Silknet.colors.backgroundLayer)
                .padding(24.dp)
        ) {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                SilknetButton("Primary", {}, variant = SilknetButtonVariant.Primary, enabled = false)
                SilknetButton("Soft", {}, variant = SilknetButtonVariant.PrimarySoft, enabled = false)
                SilknetButton("Secondary", {}, variant = SilknetButtonVariant.Secondary, enabled = false)
                SilknetButton("Ghost", {}, variant = SilknetButtonVariant.Ghost, enabled = false)
                SilknetButton("Link", {}, variant = SilknetButtonVariant.Link, enabled = false)
            }
        }
    }
}
